File: leetcode/graph/2737_Find_the_Closest_Marked_Node/codes/spfa.py
from collections import deque

class Solution:
    def minimumDistance(self, n: int, edges: list[list[int]], s: int, marked: list[int]) -> int:
        graph = [[] for _ in range(n)]
        for u, v, w in edges:
            graph[u].append((v, w))
        dist = [float('inf')] * n
        dist[s] = 0
        q = deque([s])
        in_queue = [False] * n
        in_queue[s] = True
        while q:
            u = q.popleft()
            in_queue[u] = False
            for v, w in graph[u]:
                if dist[v] > dist[u] + w:
                    dist[v] = dist[u] + w
                    if not in_queue[v]:
                        q.append(v)
                        in_queue[v] = True
        ans = min(dist[m] for m in marked)
        return -1 if ans == float('inf') else ans

File: leetcode/graph/2737_Find_the_Closest_Marked_Node/codes/spfa.cpp
#include <vector>
#include <queue>
#include <climits>
using namespace std;

class Solution {
public:
    int minimumDistance(int n, vector<vector<int>>& edges, int s, vector<int>& marked) {
        vector<vector<pair<int,int>>> graph(n);
        for (auto &e : edges) graph[e[0]].push_back({e[1], e[2]});
        vector<int> dist(n, INT_MAX);
        dist[s] = 0;
        queue<int> q;
        vector<bool> inQueue(n, false);
        q.push(s);
        inQueue[s] = true;
        while (!q.empty()) {
            int u = q.front(); q.pop();
            inQueue[u] = false;
            for (auto &[v, w] : graph[u]) {
                if (dist[v] > dist[u] + w) {
                    dist[v] = dist[u] + w;
                    if (!inQueue[v]) {
                        q.push(v);
                        inQueue[v] = true;
                    }
                }
            }
        }
        int ans = INT_MAX;
        for (int m : marked) ans = min(ans, dist[m]);
        return ans == INT_MAX ? -1 : ans;
    }
};
